Fix(#2769): Modify the annotation processor in 2.x to give a warning if a plugin builder attribute does not have a public setter.

jaykataria1111 opened the pull request:
The following commit modifies the annotation processor in 2.x to shell out a warning if the plugin builder attribute does not have a public setter. 

A ` @SuppressWarnings("log4j.public.setter")` attribute can be used to ignore this compilation warning incase it is a known issue.

## Checklist

* Base your changes on `2.x` branch if you are targeting Log4j 2; use `main` otherwise
* `./mvnw verify` succeeds (if it fails due to code formatting issues reported by Spotless, simply run `./mvnw spotless:apply` and retry)
* Non-trivial changes contain an entry file in the `src/changelog/.2.x.x` directory
* Tests for the changes are provided
* [Commits are signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) (optional, but highly recommended)

ppkarwasz commented on log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessorPublicSetterTest.java:
```diff
@@ -0,0 +1,107 @@
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.ToolProvider;
+import org.junit.Test;
```
```suggestion
import org.junit.jupiter.api.Test;
```

Please use JUnit 5. Our JUnit 4 tests are being currently rewritten to JUnit 5.

When using JUnit 5, the recommended style is to make all classes and methods package-private instead of `public`.

jaykataria1111 replied:
Done! 👍 😃

ppkarwasz commented on log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessorPublicSetterTest.java:
```diff
@@ -0,0 +1,107 @@
+        // Get the source files
+        Path sourceFile = Paths.get(FAKE_PLUGIN_CLASS_PATH);
+
+        assertThat(Files.exists(sourceFile)).isTrue();
```
```suggestion
        assertThat(sourceFile).exists();
```

ppkarwasz commented on log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessorPublicSetterTest.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @Test
+    public void IgnoreWarningWhenSuppressWarningsIsPresent() {
```
```suggestion
    public void ignoreWarningWhenSuppressWarningsIsPresent() {
```

jaykataria1111 replied:
This is was such a bad typo 😿 Thanks and fixed :)

ppkarwasz commented on log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessorPublicSetterTest.java:
```diff
@@ -0,0 +1,107 @@
+        // Get the source files
+        Path sourceFile = Paths.get(FAKE_PLUGIN_CLASS_PATH);
+        System.out.println(FAKE_PLUGIN_CLASS_PATH);
+        assertThat(Files.exists(sourceFile)).isTrue();
```
```suggestion
        assertThat(sourceFile).exists();
```

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor.java:
```diff
@@ -107,6 +120,93 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
+
+                        Types typeUtils = processingEnv.getTypeUtils();
+
+                        boolean followsNamePattern = methodName
+                                        .toLowerCase(Locale.ROOT)
+                                        .equals(String.format("set%s", fieldName.toLowerCase(Locale.ROOT)))
+                                || methodName
+                                        .toLowerCase(Locale.ROOT)
+                                        .equals(String.format("with%s", fieldName.toLowerCase(Locale.ROOT)));
+
+                        if (fieldName.toLowerCase(Locale.ROOT).startsWith("is")) {
+                            followsNamePattern = methodName
+                                            .toLowerCase(Locale.ROOT)
+                                            .equals(String.format(
+                                                    "set%s",
+                                                    fieldName
+                                                            .toLowerCase(Locale.ROOT)
+                                                            .substring(2)))
+                                    || methodName
+                                            .toLowerCase(Locale.ROOT)
+                                            .equals(String.format(
+                                                    "with%s",
+                                                    fieldName
+                                                            .toLowerCase(Locale.ROOT)
+                                                            .substring(2)));
+                        }
```
I think we could be stricter here and use a case-sensitive comparison, for example add a helper method that:

1. Removes an `is` prefix from the field name.
2. Capitalizes the first letter.
3. Compares with the method name with `set` or `with` removed.

jaykataria1111 replied:
Great Idea, I concur. 

Gotcha, we are going to get some more errors here, I am going to add more suppress warnings in them.

We should deprecate the old methods, and create new methods for setters, which have improper naming right now, what do you think, can I make an issue for this?

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor.java:
```diff
@@ -107,6 +120,93 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
+            processingEnv
+                    .getMessager()
+                    .printMessage(
+                            Diagnostic.Kind.WARNING,
```
I think that `ERROR` would be fine here.

Just make sure that the message contains a mention about the `@SuppressWarnings` annotation, so third-party Log4j plugin developers don't get crazy if they can not create a setter.

jaykataria1111 replied:
>don't get crazy if they can not create a setter.

That makes sense 🤣 ,  let me add it to the message!

ppkarwasz reviewed (changes requested):
@jaykataria1111,

Thanks for your contribution! I added some minor remarks below.

If you haven't done it already, please [sign the ASF ICLA](https://www.apache.org/licenses/contributor-agreements.html).

jaykataria1111 commented:
Sounds good @ppkarwasz let me fix and rebase this! 🙂

jaykataria1111 commented:
Hi @ppkarwasz 

>If you haven't done it already, please [sign the ASF ICLA](https://www.apache.org/licenses/contributor-agreements.html).

Thanks for sharing this and done.

Also addressed all the comments, and rebased: d6e4898

jaykataria1111 commented:
>If you haven't done it already, please [sign the ASF ICLA](https://www.apache.org/licenses/contributor-agreements.html).

Hi @ppkarwasz, thansk for sharing this and done: 

I have rebased, addressed the comments, and signed my changes here: [024654e](https://github.com/apache/logging-log4j2/pull/3195/commits/024654ee09be41f26a732e74ccaa35476936527b)

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor.java:
```diff
@@ -107,6 +123,93 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
+        }
+    }
+
+    private void processBuilderAttributes(final VariableElement element) {
```
```suggestion
    private void processBuilderAttribute(final VariableElement element) {
```

Small typo.

ppkarwasz reviewed (approved):
LGTM, thanks!

ppkarwasz commented:
@jaykataria1111,

I sent you an invitation to the Apache Slack to the address you use in your commits.

jaykataria1111 commented:
Hi @ppkarwasz the build was failing, hence I put the pr in draft 🥲. Thanks for approving it, but I will have to make a revision.

I am not sure I got the slack invitation, was the email jaykataria1111@gmail.com?

ppkarwasz commented:
No, I used the one from you University/Alma Mater that you use for your commits.

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor.java:
```diff
@@ -39,14 +41,20 @@
 import javax.tools.Diagnostic;
 import javax.tools.FileObject;
 import javax.tools.StandardLocation;
+import org.apache.commons.lang3.StringUtils;
```
This is the reason the OSGi builds fail.

We don't have `commons-lang3` as direct dependency. It is brought in the project by the **optional** `commons-compress` dependency.

BND does not automatically mark optional Maven dependencies as optional OSGi dependencies and for good reasons (see bndtools/bnd#2713 for a discussion), so the OSGi tests detected that the **required** `org.apache.commons.lang3` package is missing. Please don't use Commons Lang 3, we are a bottom-of-the-stack library with no required dependencies.

jaykataria1111 replied:
I see, yeah log4j is pretty bottom stack, we should be minimizing dependencies, thanks a lot @ppkarwasz for finding this out! I am sorry that you had to struggle for this. I have corrected it now :)

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor.java:
```diff
@@ -107,6 +123,93 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
+     * @param fieldName who's setter we are checking.
+     * @return The expected fieldName that will come after withxxxx or setxxxx
+     */
+    public static String expectedFieldNameInASetter(String fieldName) {
```
Please make this `private`, we only expose public methods, when we need to.

jaykataria1111 replied:
Done

ppkarwasz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor.java:
```diff
@@ -107,6 +123,93 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
+    public static String expectedFieldNameInASetter(String fieldName) {
+        if (fieldName.startsWith("is")) fieldName = fieldName.substring(2);
+
+        if (StringUtils.isEmpty(fieldName)) return ""; // Return empty string if the field is a blank String.
+
+        return String.format(
+                "%s%s",
+                fieldName.substring(0, 1).toUpperCase(Locale.ROOT),
+                fieldName.substring(1)); // we do not need to Lowercase here examples like field name 'ignoreExceptions'
```
```suggestion
        return fieldName.isEmpty() ? fieldName : Character.toUpperCase(fieldName.charAt(0)) + fieldName.substring(1);
```

In this case, we don't need `StringUtils.isEmpty()`, since the argument is certainly non null. In the cases we need it, we have a copy in `o.a.l.l.util.Strings.isEmpty()`.

jaykataria1111 replied:
Changed as per suggestion 😄

ppkarwasz reviewed (changes requested):
Hi @jaykataria1111,

The OSGi fail is due to an "unauthorized" usage of Commons Lang 3. All optional Maven dependencies, must be **explicitly** marked as optional in OSGi, if we need them. In this case we don't.

jaykataria1111 commented:
Hi @ppkarwasz addressed all the comments here we should be good :)

ppkarwasz reviewed (approved):
@jaykataria1111,

Thank you! I'll merge the PR, when the CI checks succeed.
